### PR TITLE
Liquidation

### PR DESCRIPTION
**Pull request by frankyin-factual**

### CODE REVIEW

**Comment by wangjj9219** [comment]
For StableAsset, no matter what the para `limit` is, swap always on ExactSupply. There may be overswap amount, and next it needs to be processed or not.

File: modules/cdp-treasury/src/lib.rs [hunk 1]
```diff
@@ -389,14 +404,71 @@ impl<T: Config> CDPTreasuryExtended<T::AccountId> for Pallet<T> {
...
+					total_supply: _,
+					redeem_amount: _,
+				} = T::StableAsset::get_redeem_proportion_amount(&yield_info, supply_limit)
+					.ok_or(Error::<T>::CannotSwap)?;
+				let mut sum = 0;
+				let mut swap_paths = vec![];
+				let mut redeem_limits = vec![];
+				let mut swap_limits = vec![];
+
+				for i in 0..amounts.len() {
```
[the remark this answers is not in the document]
Comment (frankyin-factual) [reply modules/cdp-treasury/src/lib.rs:435]: Thanks. This is fixed.

**Comment by frankyin-factual** [comment]
> For StableAsset, no matter what the para `limit` is, swap always on ExactSupply. There may be excess overswap amount, and next it needs to be processed or not. I add a commit to handle it.
> 
> Some test cases are missing and need to be filled

Tests are in `cdp-treasury` and I think I covered most cases. However, still not sure what's going in with the `evm` tests.

**Comment by frankyin-factual** [comment]
Tests fixed.

**Comment by xlc** [comment]
Can you merge master and why failed `swap_collateral_to_stable` will modify storage?

**Comment by frankyin-factual** [comment]
> swap_collateral_to_stable

I didn't test on the pallet calls but the module function, which doesn't use the `transactional` macro.

**Comment by frankyin-factual** [comment]
master merged

**Comment by wangjj9219** [comment]
> > swap_collateral_to_stable
> 
> I didn't test on the pallet calls but the module function, which doesn't use the `transactional` macro.

Because `swap_collateral_to_stable` will be called by ended auction, it's not pallet calls, need to add macro `transactional` for `swap_collateral_to_stable` if failed will modify storage

File: modules/cdp-treasury/src/lib.rs [hunk 2]
```diff
@@ -389,14 +409,68 @@ impl<T: Config> CDPTreasuryExtended<T::AccountId> for Pallet<T> {
...
+				ensure!(
+					yield_info.total_supply >= pool_info.total_supply,
+					Error::<T>::CannotSwap,
+				);
+				let RedeemProportionResult {
+					amounts,
+					balances: _,
+					fee_amount: _,
+					total_supply: _,
+					redeem_amount: _,
```
Comment (shaunxw) [inline modules/cdp-treasury/src/lib.rs:428]:
```suggestion
					..
```

File: modules/cdp-treasury/src/lib.rs [hunk 3]
```diff
@@ -389,14 +409,68 @@ impl<T: Config> CDPTreasuryExtended<T::AccountId> for Pallet<T> {
...
+				} = T::StableAsset::get_redeem_proportion_amount(&yield_info, supply_limit)
+					.ok_or(Error::<T>::CannotSwap)?;
+				let mut swap_paths = vec![];
+				let mut redeem_limits = vec![];
+				let mut swap_limits = vec![];
```
Comment (shaunxw) [inline modules/cdp-treasury/src/lib.rs:433]: Use `Vec::with_capacity(amounts.len())`

File: modules/cdp-treasury/src/lib.rs [hunk 4]
```diff
@@ -389,14 +409,68 @@ impl<T: Config> CDPTreasuryExtended<T::AccountId> for Pallet<T> {
...
+				}
+				T::StableAsset::redeem_proportion(&Self::account_id(), stable_asset_id, supply_limit, redeem_limits)?;
+
+				let mut supply_sum = 0;
+				let mut target_sum = 0;
+				for i in 0..amounts.len() {
+					let response =
+						T::DEX::swap_with_specific_path(&Self::account_id(), &swap_paths[i], swap_limits[i])?;
+					supply_sum += response.0;
+					target_sum += response.1;
```
Comment (shaunxw) [inline modules/cdp-treasury/src/lib.rs:458]: `saturating_add` or `checked_add` should be preferred

**Review by wangjj9219 — APPROVED** [review]

**Comment by wangjj9219** [comment]
LGTM

**Pull request merged**